### ISSUE TITLE
Exception is raised when association is defined in abstract class

### ISSUE THREAD

**Issue by mehanoid** [MaxLap/activerecord_where_assoc#10]
Given the following code:

```ruby
# frozen_string_literal: true

require 'bundler/inline'

gemfile(true) do
  source 'https://rubygems.org'

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem 'activerecord', '~> 7.0.0'
  gem 'sqlite3'
  # gem 'pg'
  gem 'activerecord_where_assoc', github: 'MaxLap/activerecord_where_assoc'
end

require 'active_record'
require 'minitest/autorun'
require 'logger'

ActiveRecord::Base.establish_connection(adapter: 'sqlite3', database: ':memory:')
# ActiveRecord::Base.establish_connection(adapter: 'postgresql', database: 'activerecord_where_assoc_pg')
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :posts, force: true do |t|
  end

  create_table :comments, force: true do |t|
    t.integer :commentable_id
    t.string :commentable_type
  end
end

class BasePost < ActiveRecord::Base
  self.abstract_class = true

  has_many :comments, as: :commentable
end

class Post < BasePost
end

class Comment < ActiveRecord::Base
  belongs_to :commentable, polymorphic: true
end

class BugTest < Minitest::Test
  def test_where_assoc_exists
    post = Post.create!
    post_with_comments = Post.create!
    comment = Comment.create! commentable: post_with_comments

    assert_equal Post.where_assoc_exists(:comments).to_a, [post_with_comments]
  end
end
```

Then we have the following error:
```
ActiveRecord::StatementInvalid: SQLite3::SQLException: no such column: .id
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/sqlite3-1.4.4/lib/sqlite3/database.rb:152:in `initialize'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/sqlite3-1.4.4/lib/sqlite3/database.rb:152:in `new'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/sqlite3-1.4.4/lib/sqlite3/database.rb:152:in `prepare'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/active_record/connection_adapters/sqlite3/database_statements.rb:50:in `block (2 levels) in exec_query'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activesupport-7.0.3.1/lib/active_support/concurrency/share_lock.rb:187:in `yield_shares'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activesupport-7.0.3.1/lib/active_support/dependencies/interlock.rb:41:in `permit_concurrent_loads'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/active_record/connection_adapters/sqlite3/database_statements.rb:47:in `block in exec_query'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activesupport-7.0.3.1/lib/active_support/concurrency/load_interlock_aware_monitor.rb:25:in `handle_interrupt'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activesupport-7.0.3.1/lib/active_support/concurrency/load_interlock_aware_monitor.rb:25:in `block in synchronize'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activesupport-7.0.3.1/lib/active_support/concurrency/load_interlock_aware_monitor.rb:21:in `handle_interrupt'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activesupport-7.0.3.1/lib/active_support/concurrency/load_interlock_aware_monitor.rb:21:in `synchronize'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/active_record/connection_adapters/abstract_adapter.rb:765:in `block in log'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activesupport-7.0.3.1/lib/active_support/notifications/instrumenter.rb:24:in `instrument'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/active_record/connection_adapters/abstract_adapter.rb:756:in `log'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/active_record/connection_adapters/sqlite3/database_statements.rb:46:in `exec_query'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/active_record/connection_adapters/abstract/database_statements.rb:560:in `select'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/active_record/connection_adapters/abstract/database_statements.rb:66:in `select_all'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/active_record/connection_adapters/abstract/query_cache.rb:110:in `select_all'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/active_record/querying.rb:54:in `_query_by_sql'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/active_record/relation.rb:941:in `block in exec_main_query'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/active_record/relation.rb:961:in `skip_query_cache_if_necessary'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/active_record/relation.rb:927:in `exec_main_query'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/active_record/relation.rb:913:in `block in exec_queries'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/active_record/relation.rb:961:in `skip_query_cache_if_necessary'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/active_record/relation.rb:907:in `exec_queries'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/active_record/relation.rb:695:in `load'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/active_record/relation.rb:250:in `records'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/active_record/relation.rb:245:in `to_ary'
    activerecord_where_assoc.rb:53:in `test_where_assoc_exists'
```

And with postgres adapter there is another error:

```
TypeError: no implicit conversion of nil into String
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/active_record/connection_adapters/postgresql/utils.rb:26:in `quote_ident'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/active_record/connection_adapters/postgresql/utils.rb:26:in `quoted'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/active_record/connection_adapters/postgresql/quoting.rb:58:in `quote_table_name'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/arel/visitors/to_sql.rb:801:in `quote_table_name'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/arel/visitors/to_sql.rb:734:in `visit_Arel_Attributes_Attribute'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/arel/visitors/visitor.rb:30:in `visit'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/arel/visitors/to_sql.rb:654:in `visit_Arel_Nodes_Equality'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/arel/visitors/visitor.rb:30:in `visit'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/arel/visitors/to_sql.rb:827:in `block in inject_join'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/arel/visitors/to_sql.rb:825:in `each'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/arel/visitors/to_sql.rb:825:in `each_with_index'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/arel/visitors/to_sql.rb:825:in `inject_join'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/arel/visitors/to_sql.rb:612:in `visit_Arel_Nodes_And'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/arel/visitors/visitor.rb:30:in `visit'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/arel/visitors/to_sql.rb:827:in `block in inject_join'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/arel/visitors/to_sql.rb:825:in `each'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/arel/visitors/to_sql.rb:825:in `each_with_index'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/arel/visitors/to_sql.rb:825:in `inject_join'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/arel/visitors/to_sql.rb:179:in `collect_nodes_for'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/arel/visitors/to_sql.rb:159:in `visit_Arel_Nodes_SelectCore'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/arel/visitors/to_sql.rb:124:in `block in visit_Arel_Nodes_SelectStatement'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/arel/visitors/to_sql.rb:123:in `each'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/arel/visitors/to_sql.rb:123:in `inject'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/arel/visitors/to_sql.rb:123:in `visit_Arel_Nodes_SelectStatement'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/arel/visitors/visitor.rb:30:in `visit'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/arel/visitors/visitor.rb:11:in `accept'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/arel/visitors/to_sql.rb:18:in `compile'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/active_record/connection_adapters/abstract/database_statements.rb:37:in `to_sql_and_binds'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/active_record/connection_adapters/abstract/database_statements.rb:13:in `to_sql'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/active_record/relation.rb:731:in `block in to_sql'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/active_record/connection_adapters/abstract_adapter.rb:285:in `unprepared_statement'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/activerecord-7.0.3.1/lib/active_record/relation.rb:731:in `to_sql'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/bundler/gems/activerecord_where_assoc-1da9429058ce/lib/active_record_where_assoc/core_logic.rb:17:in `block in sql_for_any_exists'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/bundler/gems/activerecord_where_assoc-1da9429058ce/lib/active_record_where_assoc/core_logic.rb:17:in `map'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/bundler/gems/activerecord_where_assoc-1da9429058ce/lib/active_record_where_assoc/core_logic.rb:17:in `sql_for_any_exists'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/bundler/gems/activerecord_where_assoc-1da9429058ce/lib/active_record_where_assoc/core_logic.rb:65:in `assoc_exists_sql'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/bundler/gems/activerecord_where_assoc-1da9429058ce/lib/active_record_where_assoc/relation_returning_methods.rb:256:in `where_assoc_exists'
    /home/meh/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/bundler/gems/activerecord_where_assoc-1da9429058ce/lib/active_record_where_assoc/relation_returning_delegates.rb:10:in `where_assoc_exists'
    activerecord_where_assoc.rb:53:in `test_where_assoc_exists'
```

**Comment by MaxLap** [comment]
Nice bug, and nice bug report. Didn't think of doing any tests with abstract models, even less so where the association is on the abstract model.

I got a probable fix for it, but I still need to write tests to avoid possible regressions and to validate against other cases with abstract models.

**Comment by MaxLap** [comment]
I released 1.1.3 which fixes this.

Thanks again for reporting the bug.

**Comment by mehanoid** [comment]
It works now, thank you for the quick fix! :+1:
